### PR TITLE
ARROW-3180: [C++] Add docker-compose setup to simulate Travis CI run locally

### PR DESCRIPTION
**Pull request by kszucs**
Also resolves: 

> ARROW-3078: [Python] Docker integration tests should not contaminate the local Python development environment

Working:
- [x] rust
- [x] go
- [x] c_glib
- [x] ruby
- [x] java
- [x] cpp
- [x] python
- [x] hdfs integration

The rest:
- js (gulp error...)
- matlab
- R
- dask integration
- spark integration
- gen_apidocs
- hiveserver2
- iwyu

Perhaps resolve them in follow-up PRs.

### CODE REVIEW

**Comment by kszucs** [comment]
@wesm could You take a look at this docker-compose setup? 

Executing `python` tests: `docker-compose run python` 
Executing `cpp` tests: `docker-compose run cpp` 
Executing `c_glib` tests: `docker-compose run c_glib` 
Executing `hdfs` tests: `docker-compose run hdfs-integration` 
etc.

It's also possible to mount arrow externally for an editable setup:
```bash
docker-compose run -v local/path/to/arrow:/arrow python
```

Ideally "the rest" of the containers should be ported too.

**Comment by wesm** [comment]
Seems like having a more accessible pre-commit hook could help. I run the following code before every push to github

```
function arrow_preflight {
    ARROW_PREFLIGHT_DIR=$ARROW_ROOT/cpp/preflight-build
    mkdir -p $ARROW_PREFLIGHT_DIR
    pushd $ARROW_ROOT/cpp
    python build-support/lint_cpp_cli.py src || return
    popd
    pushd $ARROW_PREFLIGHT_DIR
    arrow_cmake
    ninja format
    ninja lint || return
    popd
    pushd $ARROW_ROOT/python
    flake8 --count pyarrow || return
    flake8 --count --config=.flake8.cython pyarrow || return
    popd
}
```

**Comment by wesm** [comment]
@kszucs thanks a lot for working on this. I will take it for a spin when I can and report back feedback

**Comment by kszucs** [comment]
Sadly we can't chain the builds to depend on each other purely with docker compose e.g.: `conda -> cpp -> c_glib -> ruby`, so chaining with a Makefile would improve the build times.

**Comment by kszucs** [comment]
TODO: crossbow should have it's own image

**Comment by wesm** [comment]
I'm working on code reviews, but hope to get to this soon

**Review by wesm — APPROVED** [review]
+1, thanks, this is good progress modularizing these dockerfiles! I think the hope of getting everything 100% working on one patch is a lot to ask, but we can work toward it in follow-up PRs. Will merge once the build passes

**Comment by xhochy** [comment]
We should add a small word about these files to the docs.

**Comment by wesm** [comment]
He mentioned that the JS build is broken in the PR description

**Comment by kszucs** [comment]
Well, I'm a JS noob :) 

Strange that the js build works outside of the docker container, I've tried to tune the dockerignore file to include all required packaging files in the image, but sadly with no luck.

Created issues for the follow-up tasks:
- Documentation https://issues.apache.org/jira/browse/ARROW-3365
- JS Dockerfile https://issues.apache.org/jira/browse/ARROW-3365
- R Dockerfile https://issues.apache.org/jira/browse/ARROW-3366
- Spark integration https://issues.apache.org/jira/browse/ARROW-3367
- Dask integration https://issues.apache.org/jira/browse/ARROW-3368

**Comment by kszucs** [comment]
It indeed ignores shm_size. What's your docker engine? Are You using docker on mac or docker-machine?

**Comment by xhochy** [comment]
Docker on mac: 18.06.1-ce-mac73

**Comment by xhochy** [comment]
When running the manylinux images using `docker run --shm-size…` it works.

**Comment by kszucs** [comment]
Same for me, i can be a docker on mac issue. I'm spinning up a linux box...

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+  js:
+    image: arrow:js
+    build:
+      context: .
+      dockerfile: js/Dockerfile
+
+  python:
+    image: arrow:python-${PYTHON_VERSION:-3.6}
+    shm_size: 2G
```
Comment (kszucs) [inline docker-compose.yml]: @xhochy docker has a "great" documentation

Comment (xhochy) [reply docker-compose.yml]: 🙈

**Review by xhochy — APPROVED** [review]
+1, Python tests now also pass for me.

**Comment by kszucs** [comment]
@xhochy CI error is related to the wheel issue

**Comment by wesm** [comment]
Thanks everyone. This will help a lot

**Pull request closed without merging**